core/src/main/java/hudson/Util.java

```diff
@@ -512,34 +287,33 @@ private static String deleteFailExceptionMessage(File whatWeWereTryingToRemove,
      * Checks if the given file represents a symlink.
      */
     public static boolean isSymlink(@Nonnull File file) throws IOException {
-        /*
-         *  Windows Directory Junctions are effectively the same as Linux symlinks to directories.
-         *  Unfortunately, the Java 7 NIO2 API function isSymbolicLink does not treat them as such.
-         *  It thinks of them as normal directories.  To use the NIO2 API & treat it like a symlink,
-         *  you have to go through BasicFileAttributes and do the following check:
-         *     isSymbolicLink() || isOther()
-         *  The isOther() call will include Windows reparse points, of which a directory junction is.
-         *  It also includes includes devices, but reading the attributes of a device with NIO fails
-         *  or returns false for isOther(). (i.e. named pipes such as \\.\pipe\JenkinsTestPipe return
-         *  false for isOther(), and drives such as \\.\PhysicalDrive0 throw an exception when
-         *  calling readAttributes.
-         */
-        try {
-            Path path = fileToPath(file);
-            BasicFileAttributes attrs = Files.readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
-            if (attrs.isSymbolicLink()) {
-                return true;
-            } else if (attrs instanceof DosFileAttributes) {
-                /* Returns true for non-symbolic link reparse points and devices. We could call
-                 * WindowsFileAttributes#isReparsePoint with reflection instead to exclude devices,
-                 * but as mentioned in the above comment this does not appear to be an issue.
-                 */
-                return attrs.isOther();
-            } else {
+        return isSymlink(fileToPath(file));
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static boolean isSymlink(@Nonnull Path path) {
+        if (Functions.isWindows()) {
+            /*
+             *  Windows Directory Junctions are effectively the same as Linux symlinks to directories.
+             *  Unfortunately, the Java 7 NIO2 API function isSymbolicLink does not treat them as such.
+             *  It thinks of them as normal directories.  To use the NIO2 API & treat it like a symlink,
+             *  you have to go through BasicFileAttributes and do the following check:
+             *     isSymbolicLink() || isOther()
+             *  The isOther() call will include Windows reparse points, of which a directory junction is.
+             *  It also includes includes devices, but reading the attributes of a device with NIO fails
+             *  or returns false for isOther(). (i.e. named pipes such as \\.\pipe\JenkinsTestPipe return
+             *  false for isOther(), and drives such as \\.\PhysicalDrive0 throw an exception when
+             *  calling readAttributes.
+             */
+            try {
+                return path.compareTo(path.toRealPath()) != 0;
```

Comment:
Using `Files.isSymbolicLink()` works on Windows only for literal symbolic links. Directory junctions in Windows are apparently very similar to symbolic links, but they have slightly different rules about permissions required to create them (junctions don't require admin permissions or GPO changes) as well as which strategy is used to resolve remote directories in the junction.

What's neat here is that we can effectively check if a file is some sort of logical symbolic link regardless of its file attributes. Calling `toRealPath()` without specifying `NOFOLLOW_LINKS` means it will resolve any sort of link (FS-specific of course). This has the added interesting effect of considering paths like `..` to be a symbolic link which it sort of is.

And yes, the comment is now outdated. I'll update that with proper info.